streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java

```diff
@@ -0,0 +1,160 @@
+
+    KStreamSessionWindowAggregate(final SessionWindows windows,
+                                  final String storeName,
+                                  final Initializer<T> initializer,
+                                  final Aggregator<? super K, ? super V, T> aggregator,
+                                  final Merger<? super K, T> sessionMerger) {
+        this.windows = windows;
+        this.storeName = storeName;
+        this.initializer = initializer;
+        this.aggregator = aggregator;
+        this.sessionMerger = sessionMerger;
+    }
+
+    @Override
+    public Processor<K, V> get() {
+        return new KStreamSessionWindowAggregateProcessor();
+    }
+
+    @Override
+    public void enableSendingOldValues() {
+        sendOldValues = true;
+    }
+
+    private class KStreamSessionWindowAggregateProcessor extends AbstractProcessor<K, V> {
+
+        private SessionStore<K, T> store;
+        private TupleForwarder<Windowed<K>, T> tupleForwarder;
+
+        @SuppressWarnings("unchecked")
+        @Override
+        public void init(ProcessorContext context) {
+            super.init(context);
+            store = (SessionStore<K,  T>) context.getStateStore(storeName);
+            tupleForwarder = new TupleForwarder<>(store, context, new ForwardingCacheFlushListener<K, V>(context, sendOldValues), sendOldValues);
+        }
+
+        @Override
+        public void process(final K key, final V value) {
+            // if the key is null, we do not need proceed aggregating
+            // the record with the table
+            if (key == null) {
+                return;
+            }
+
+            final long timestamp = context().timestamp();
+            final List<KeyValue<Windowed<K>, T>> merged = new ArrayList<>();
+            final TimeWindow newTimeWindow = new TimeWindow(timestamp, timestamp);
+            TimeWindow mergedWindow = newTimeWindow;
+            T agg = initializer.apply();
+
+            try (final KeyValueIterator<Windowed<K>, T> iterator = store.findSessionsToMerge(key, timestamp - windows.inactivityGap(),
+                                                                                               timestamp + windows.inactivityGap())) {
+                while (iterator.hasNext()) {
+                    final KeyValue<Windowed<K>, T> next = iterator.next();
+                    merged.add(next);
+                    agg = sessionMerger.apply(key, agg, next.value);
+                    mergedWindow = mergeTimeWindow(mergedWindow, (TimeWindow) next.key.window());
+                }
+            }
+
+            agg = aggregator.apply(key, value, agg);
+            final Windowed<K> sessionKey = new Windowed<>(key, mergedWindow);
+            if (!mergedWindow.equals(newTimeWindow)) {
+                for (final KeyValue<Windowed<K>, T> session : merged) {
+                    store.remove(session.key);
+                    tupleForwarder.maybeForward(session.key, null, session.value);
```

Comment:
We discussed that at length during the design phase. We decided that we'd stick with the same forwarding as we do with everything else as it is consistent. 
It is actually listed as a rejected alternative in the KIP